### PR TITLE
Stability fixes

### PR DESCRIPTION
**Pull request by sargun**

### CODE REVIEW

File: records/generator.go [hunk 1]
```diff
@@ -286,12 +284,20 @@ func (rg *RecordGenerator) loadWrap(ip, port string) (state.State, error) {
 	if err != nil {
 		return state.State{}, err
 	}
-	if rip := leaderIP(sj.Leader); rip != ip {
-		logging.VeryVerbose.Println("Warning: master changed to " + ip)
-		sj, err = rg.loadFromMaster(rip, port)
-		return sj, err
+	if sj.Leader != "" {
+		// FIXME(prozlach): Still, we are assuming that sj.Leader is a correct
```
Comment (sargun) [inline records/generator.go]: I think we'll crash here? Right?

**Comment by sargun** [comment]
Is there any way we can add tests to confirm this robustness? @drewkerrigan

**Comment by drewkerrigan** [comment]
@sargun I added some simple validation in records.leaderIP and added an error to the return. Invalid states should now return an error back to the caller to deal with.

**Comment by sargun** [comment]
LGTM. @jdef

File: records/generator.go [hunk 2]
```diff
@@ -653,9 +662,17 @@ func (rg *RecordGenerator) insertRR(name, host string, kind rrsKind) (added bool
...
-func leaderIP(leader string) string {
-	pair := strings.Split(leader, "@")[1]
-	return strings.Split(pair, ":")[0]
+func leaderIP(leader string) (string, error) {
+	nameAddressPair := strings.Split(leader, "@")
+	if len(nameAddressPair) != 2 {
+		return "", errors.New("Invalid leader address: " + leader)
+	}
+	hostPort := nameAddressPair[1]
+	hostPortPair := strings.Split(hostPort, ":")
```
Comment (jdef) [inline records/generator.go]: use net.SplitHostPort

File: records/generator.go [hunk 3]
```diff
@@ -286,12 +284,23 @@ func (rg *RecordGenerator) loadWrap(ip, port string) (state.State, error) {
...
+		}
+		if rip != ip {
+			logging.VeryVerbose.Println("Warning: master changed to " + ip)
+			sj, err = rg.loadFromMaster(rip, port)
+			return sj, err
+		}
+		return sj, nil
 	}
-	return sj, nil
+	msg := "Fetched state.json does not contain leader information"
```
Comment (jdef) [inline records/generator.go]: why is msg a separate var here?

File: records/generator.go [hunk 4]
```diff
@@ -286,12 +284,23 @@ func (rg *RecordGenerator) loadWrap(ip, port string) (state.State, error) {
...
+	if sj.Leader != "" {
+		var rip string
+
+		rip, err = leaderIP(sj.Leader)
+		if err != nil {
+			return sj, err
+		}
+		if rip != ip {
+			logging.VeryVerbose.Println("Warning: master changed to " + ip)
+			sj, err = rg.loadFromMaster(rip, port)
```
Comment (jdef) [inline records/generator.go]: why assign, then return? should just `return rg.loadFromMaster(...)`

**Comment by jdef** [comment]
added some nit-pick comments. look similar to another PR that I saw a while back. if this is just an updated version of that PR, would you please xref?

**Comment by drewkerrigan** [comment]
@jdef I've addressed your comments, let me know if you see anything else. The PR you're referring to is: #412 - this is an updated version.

**Comment by sargun** [comment]
@jdef Github refused to update the diff correctly on #412.

File: records/generator.go [hunk 5]
```diff
@@ -170,7 +170,7 @@ func (rg *RecordGenerator) ParseState(c Config, masters ...string) error {
 	// find master -- return if error
 	sj, err := rg.findMaster(masters...)
 	if err != nil {
-		logging.Error.Println("no master")
+		logging.Error.Println("Failed to fetch state.json")
```
Comment (sargun) [inline records/generator.go]: Why don't we print the error here, when on line 210 we print the error?

File: records/generator.go [hunk 6]
```diff
@@ -203,36 +203,34 @@ func (rg *RecordGenerator) findMaster(masters ...string) (state.State, error) {
...
-			return sj, nil
+		if sj, err = rg.loadWrap(ip, port); err != nil {
+			logging.Error.Println("Failed to fetch state.json - trying next one. Error: ", err)
+			continue
 		}
+		return sj, nil
 	}
 
-	return sj, errors.New("no master")
+	return sj, errors.New("no more masters eligible for state.json query")
```
Comment (sargun) [inline records/generator.go]: ^No

File: records/generator.go [hunk 7]
```diff
@@ -286,12 +284,21 @@ func (rg *RecordGenerator) loadWrap(ip, port string) (state.State, error) {
 	if err != nil {
 		return state.State{}, err
 	}
-	if rip := leaderIP(sj.Leader); rip != ip {
-		logging.VeryVerbose.Println("Warning: master changed to " + ip)
-		sj, err = rg.loadFromMaster(rip, port)
-		return sj, err
+	if sj.Leader != "" {
+		var rip string
```
Comment (sargun) [inline records/generator.go]: What's a RIP?

**Comment by sargun** [comment]
A couple more nits.

**Comment by drewkerrigan** [comment]
@sargun nits addressed, squash?

**Comment by jdef** [comment]
lgtm, please squash

**Pull request merged**